provenance-verifier/pkg/provenance.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if len(certs) != 1 {
+		return nil, errors.New("unexpected number of cert pem tlog entry")
+	}
+
+	return certs[0], err
+}
+
+// FindSigningCertificate finds and verifies a matching signing certificate from a list of Rekor entry UUIDs.
+func FindSigningCertificate(ctx context.Context, uuids []string, dssePayload dsselib.Envelope, rClient *client.Rekor) (*x509.Certificate, error) {
+	attBytes, err := cjson.MarshalCanonical(dssePayload)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate through each matching UUID and perform:
+	//   * Verify TLOG entry (inclusion and signed entry timestamp against Rekor pubkey).
+	//   * Verify the signing certificate against the Fulcio root CA.
+	//   * Verify dsse envelope signature against signing certificate.
+	//   * Check signature expiration against IntegratedTime in entry.
+	//   * If all succeed, return the signing certificate.
+	for _, uuid := range uuids {
+		entry, err := verifyTlogEntry(ctx, rClient, uuid)
+		if err != nil {
+			continue
+		}
+		cert, err := extractCert(entry)
+		if err != nil {
+			continue
+		}
+		co := &cosign.CheckOpts{
+			RootCerts:      fulcio.GetRoots(),
+			CertOidcIssuer: certOidcIssuer,
+		}
+		verifier, err := cosign.ValidateAndUnpackCert(cert, co)
+		if err != nil {
+			continue
+		}
+		verifier = dsse.WrapVerifier(verifier)
+		if err := verifier.VerifySignature(bytes.NewReader(attBytes), bytes.NewReader(attBytes)); err != nil {
+			continue
+		}
+		it := time.Unix(*entry.IntegratedTime, 0)
+		if err := cosign.CheckExpiry(cert, it); err != nil {
+			continue
+		}
+		// success!
+		fmt.Printf("Verified against tlog entry %d\n", *entry.LogIndex)
+		return cert, nil
+	}
+
+	return nil, errors.New("could not find a matching signature entry")
+}
+
+func getExtension(cert *x509.Certificate, oid string) string {
+	for _, ext := range cert.Extensions {
+		if strings.Contains(ext.Id.String(), oid) {
+			return string(ext.Value)
+		}
+	}
+	return ""
+}
+
+// GetWorkflowFromCertificate gets the workflow content from the Fulcio authenticated content.
+func GetWorkflowFromCertificate(cert *x509.Certificate) (*github.RepositoryContent, error) {
+	if len(cert.URIs) == 0 {
+		return nil, errors.New("missing URI information from certificate")
+	}
+
+	// cert URI path is /org/repo/path/to/workflow@ref
+	pathParts := strings.SplitN(cert.URIs[0].Path, "/", 4)
+	if len(pathParts) < 3 {
+		return nil, errors.New("malformed URI for workflow")
+	}
+	org := pathParts[1]
+	repo := pathParts[2]
+
+	filecontent := strings.SplitN(pathParts[3], "@", 2)
+	if len(filecontent) < 2 {
+		return nil, errors.New("malformed filepath and ref for workflow content")
+	}
+
+	// Checkout the workflow path at the commit hash from the cert.
+	ctx := context.Background()
+	client := github.NewClient(nil)
+	workflowContent, _, _, err := client.Repositories.GetContents(ctx, org, repo, filecontent[0], &github.RepositoryContentGetOptions{Ref: filecontent[1]})
```

Comment:
INCORRECT! MAKE SURE I PULL THE CALLER WORKFLOW! using `1.3.6.1.4.1.57264.1.3` hash and `1.3.6.1.4.1.57264.1.3` org/repo